-1

**Failed Tests**:   Build 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-db5c7a/30961/summary.html
**COMMIT**: 432c76c5d9fbeafdc0f2ce86501ce5e275f6a9d7
**CMSSW**: CMSSW_13_1_X_2023-02-28-2300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/40895/30961/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Build

I found compilation error when building: 

<pre>>> Compiling edm plugin /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_13_1_X_2023-02-28-2300/src/DQM/CTPPS/plugins/TotemT2DQMSource.cc
>> Compiling edm plugin /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_13_1_X_2023-02-28-2300/src/DQM/CTPPS/plugins/TotemTimingDQMSource.cc
>> Building edm plugin tmp/el8_amd64_gcc11/src/DQM/CTPPS/plugins/DQMCTPPSPlugins/libDQMCTPPSPlugins.so
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02774/el8_amd64_gcc11/external/gcc/11.2.1-f9b9dfdd886f71cd63f5538223d8f161/bin/../lib/gcc/x86_64-redhat-linux-gnu/11.2.1/../../../../x86_64-redhat-linux-gnu/bin/ld: tmp/el8_amd64_gcc11/src/DQM/CTPPS/plugins/DQMCTPPSPlugins/ccGZQfoY.ltrans3.ltrans.o: in function `CTPPSRandomDQMSource::CTPPSRandomDQMSource(edm::ParameterSet const&)':
<artificial>:(.text+0x16ec): undefined reference to `vtable for CTPPSRandomDQMSource'
collect2: error: ld returned 1 exit status
gmake: *** [tmp/el8_amd64_gcc11/src/DQM/CTPPS/plugins/DQMCTPPSPlugins/libDQMCTPPSPlugins.so] Error 1
Leaving library rule at src/DQM/CTPPS/plugins
Entering library rule at DQM/CTPPS
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_13_1_X_2023-02-28-2300/src/DQM/CTPPS/src/TotemT2Segmentation.cc
>> Building shared library tmp/el8_amd64_gcc11/src/DQM/CTPPS/src/DQMCTPPS/libDQMCTPPS.so
</pre>

--- a/DQM/CTPPS/plugins/CTPPSRandomDQMSource.cc
+++ b/DQM/CTPPS/plugins/CTPPSRandomDQMSource.cc
@@ -33,128 +33,124 @@
 class CTPPSRandomDQMSource : public DQMEDAnalyzer {
 public:
   CTPPSRandomDQMSource(const edm::ParameterSet &ps);
-  ~CTPPSRandomDQMSource() override;
+  ~CTPPSRandomDQMSource() override = default;
+  static void fillDescriptions(edm::ConfigurationDescriptions &descriptions);
 
 protected:
-  void dqmBeginRun(edm::Run const &, edm::EventSetup const &) override;
   void bookHistograms(DQMStore::IBooker &, edm::Run const &, edm::EventSetup const &) override;
   void analyze(edm::Event const &e, edm::EventSetup const &eSetup) override;
 
 private:
-  edm::EDGetTokenT<edm::DetSetVector<CTPPSPixelDigi>> tokenDigi;
+  edm::EDGetTokenT<edm::DetSetVector<CTPPSPixelDigi>> const tokenDigi_;
 
-  static constexpr int NArms = 2;
-  static constexpr int NStationMAX = 3;  // in an arm
-  static constexpr int NRPotsMAX = 6;    // per station
-  static constexpr int NplaneMAX = 6;    // per RPot
-  static constexpr int RPn_first = 3, RPn_last = 4;
-  static constexpr int StationIDMAX = 4;  // possible range of ID
-  static constexpr int RPotsIDMAX = 8;    // possible range of ID
+  static constexpr int kNArms_ = 2;
+  static constexpr int kNStationMAX_ = 3;  // in an arm
+  static constexpr int kNRPotsMAX_ = 6;    // per station
+  static constexpr int kNplaneMAX_ = 6;    // per RPot
+  static constexpr int kFirstRPn_ = 3, kLastRPn_ = 4;
+  static constexpr int kStationIDMAX_ = 4;  // possible range of ID
+  static constexpr int kRPotsIDMAX_ = 8;    // possible range of ID
 
-  unsigned int rpStatusWord = 0x8008;      // 220_fr_hr(stn2rp3)+ 210_fr_hr
-  int RPstatus[StationIDMAX][RPotsIDMAX];  // symmetric in both arms
-  int StationStatus[StationIDMAX];         // symmetric in both arms
-  const int IndexNotValid = 0;
+  const std::string folderName_ = "CTPPS/RandomPixel";
 
-  MonitorElement *hBX;
+  unsigned int rpStatusWord_ = 0x8008;          // 220_fr_hr(stn2rp3)+ 210_fr_hr
+  int rpStatus_[kStationIDMAX_][kRPotsIDMAX_];  // symmetric in both arms
+  int stationStatus_[kStationIDMAX_];           // symmetric in both arms
+  const int kIndexNotValid = 0;
 
-  static constexpr int RPotsTotalNumber = NArms * NStationMAX * NRPotsMAX;
+  MonitorElement *hBX_;
 
-  int RPindexValid[RPotsTotalNumber];
-  MonitorElement *h2HitsVsBXRandoms[RPotsTotalNumber];
+  static constexpr int kRPotsTotalNumber_ = kNArms_ * kNStationMAX_ * kNRPotsMAX_;
 
-  int getRPindex(int arm, int station, int rp) {
+  int RPindexValid_[kRPotsTotalNumber_];
+  MonitorElement *h2HitsVsBXRandoms_[kRPotsTotalNumber_];
+
+  int getRPindex(int arm, int station, int rp) const {
     if (arm < 0 || station < 0 || rp < 0)
-      return (IndexNotValid);
-    if (arm > 1 || station >= NStationMAX || rp >= NRPotsMAX)
-      return (IndexNotValid);
-    int rc = (arm * NStationMAX + station) * NRPotsMAX + rp;
+      return (kIndexNotValid);
+    if (arm > 1 || station >= kNStationMAX_ || rp >= kNRPotsMAX_)
+      return (kIndexNotValid);
+    int rc = (arm * kNStationMAX_ + station) * kNRPotsMAX_ + rp;
     return (rc);
   }
 };
 
-//----------------------------------------------------------------------------------
-
-using namespace std;
-using namespace edm;
-
 //-------------------------------------------------------------------------------
 
 CTPPSRandomDQMSource::CTPPSRandomDQMSource(const edm::ParameterSet &ps)
-    : rpStatusWord(ps.getUntrackedParameter<unsigned int>("RPStatusWord", 0x8008)) {
-  tokenDigi = consumes<DetSetVector<CTPPSPixelDigi>>(ps.getUntrackedParameter<edm::InputTag>("tagRPixDigi"));
+    : tokenDigi_(consumes<edm::DetSetVector<CTPPSPixelDigi>>(ps.getUntrackedParameter<edm::InputTag>("tagRPixDigi"))),
+      folderName_(ps.getUntrackedParameter<std::string>("folderName", "CTPPS/RandomPixel")),
+      rpStatusWord_(ps.getUntrackedParameter<unsigned int>("RPStatusWord", 0x8008)) {
+  for (int stn = 0; stn < kStationIDMAX_; stn++) {
+    stationStatus_[stn] = 0;
+    for (int rp = 0; rp < kRPotsIDMAX_; rp++)
+      rpStatus_[stn][rp] = 0;
+  }
+
+  unsigned int rpSts = rpStatusWord_ << 1;
+  for (int stn = 0; stn < kNStationMAX_; stn++) {
+    int stns = 0;
+    for (int rp = 0; rp < kNRPotsMAX_; rp++) {
+      rpSts = (rpSts >> 1);
+      rpStatus_[stn][rp] = rpSts & 1;
+      if (rpStatus_[stn][rp] > 0)
+        stns = 1;
+    }
+    stationStatus_[stn] = stns;
+  }
+
+  for (int index = 0; index < 2 * 3 * kNRPotsMAX_; index++)
+    RPindexValid_[index] = 0;
 }
 
 //--------------------------------------------------------------------------
 
-void CTPPSRandomDQMSource::dqmBeginRun(edm::Run const &run, edm::EventSetup const &) {
-  for (int stn = 0; stn < StationIDMAX; stn++) {
-    StationStatus[stn] = 0;
-    for (int rp = 0; rp < RPotsIDMAX; rp++)
-      RPstatus[stn][rp] = 0;
-  }
-
-  unsigned int rpSts = rpStatusWord << 1;
-  for (int stn = 0; stn < NStationMAX; stn++) {
-    int stns = 0;
-    for (int rp = 0; rp < NRPotsMAX; rp++) {
-      rpSts = (rpSts >> 1);
-      RPstatus[stn][rp] = rpSts & 1;
-      if (RPstatus[stn][rp] > 0)
-        stns = 1;
-    }
-    StationStatus[stn] = stns;
-  }
-
-  for (int index = 0; index < 2 * 3 * NRPotsMAX; index++)
-    RPindexValid[index] = 0;
-}
-
-//-------------------------------------------------------------------------------------
-
 void CTPPSRandomDQMSource::bookHistograms(DQMStore::IBooker &ibooker, edm::Run const &, edm::EventSetup const &) {
   ibooker.cd();
-  ibooker.setCurrentFolder("CTPPS/TrackingPixel");
+  ibooker.setCurrentFolder(folderName_);
 
-  hBX = ibooker.book1D("events per BX", "ctpps_pixel;Event.BX", 4002, -1.5, 4000. + 0.5);
+  hBX_ = ibooker.book1D("events per BX", "ctpps_pixel;Event.BX", 4002, -1.5, 4000. + 0.5);
 
-  for (int arm = 0; arm < NArms; arm++) {
+  for (int arm = 0; arm < kNArms_; arm++) {
     CTPPSDetId ID(CTPPSDetId::sdTrackingPixel, arm, 0);
-    string sd;
-    ID.armName(sd, CTPPSDetId::nPath);
+    std::string sd;
+    ID.armName(sd, CTPPSDetId::nShort);
+    sd = folderName_ + "/sector " + sd;
 
     ibooker.setCurrentFolder(sd);
 
-    for (int stn = 0; stn < NStationMAX; stn++) {
-      if (StationStatus[stn] == 0)
+    for (int stn = 0; stn < kNStationMAX_; stn++) {
+      if (stationStatus_[stn] == 0)
         continue;
       ID.setStation(stn);
-      string stnd;
-      CTPPSDetId(ID.stationId()).stationName(stnd, CTPPSDetId::nPath);
+      std::string stnd;
+      CTPPSDetId(ID.stationId()).stationName(stnd, CTPPSDetId::nShort);
+      stnd = sd + "/station " + stnd;
 
       ibooker.setCurrentFolder(stnd);
 
-      for (int rp = RPn_first; rp < RPn_last; rp++) {  // only installed pixel pots
+      for (int rp = kFirstRPn_; rp < kLastRPn_; rp++) {  // only installed pixel pots
         ID.setRP(rp);
-        string rpd, rpTitle;
+        std::string rpd, rpTitle;
         CTPPSDetId(ID.rpId()).rpName(rpTitle, CTPPSDetId::nFull);
-        CTPPSDetId(ID.rpId()).rpName(rpd, CTPPSDetId::nPath);
+        CTPPSDetId(ID.rpId()).rpName(rpd, CTPPSDetId::nShort);
+        rpd = stnd + "/" + rpd;
 
         ibooker.setCurrentFolder(rpd);
 
         int indexP = getRPindex(arm, stn, rp);
-        RPindexValid[indexP] = 1;
+        RPindexValid_[indexP] = 1;
 
-        h2HitsVsBXRandoms[indexP] = ibooker.book2D("Digi per plane per BX - random triggers",
-                                                   rpTitle + ";Event.BX;Plane",
-                                                   4002,
-                                                   -1.5,
-                                                   4000. + 0.5,
-                                                   NplaneMAX,
-                                                   0,
-                                                   NplaneMAX);
+        h2HitsVsBXRandoms_[indexP] = ibooker.book2D("Digi per plane per BX - random triggers",
+                                                    rpTitle + ";Event.BX;Plane",
+                                                    4002,
+                                                    -1.5,
+                                                    4000. + 0.5,
+                                                    kNplaneMAX_,
+                                                    0,
+                                                    kNplaneMAX_);
 
-      }  // end for(int rp=0; rp<NRPotsMAX;...
+      }  // end for(int rp=0; rp<kNRPotsMAX_;...
     }    // end of for(int stn=0; stn<
   }      // end of for(int arm=0; arm<2;...
 
@@ -164,37 +160,45 @@ void CTPPSRandomDQMSource::bookHistograms(DQMStore::IBooker &ibooker, edm::Run c
 //-------------------------------------------------------------------------------
 
 void CTPPSRandomDQMSource::analyze(edm::Event const &event, edm::EventSetup const &eventSetup) {
-  Handle<DetSetVector<CTPPSPixelDigi>> pixDigi;
-  event.getByToken(tokenDigi, pixDigi);
+  auto const pixDigi = event.getHandle(tokenDigi_);
 
-  cout << "Before checking digi" << endl;
   if (!pixDigi.isValid())
     return;
-  cout << "After checking digi" << endl;
 
-  hBX->Fill(event.bunchCrossing());
+  hBX_->Fill(event.bunchCrossing());
 
   for (int arm = 0; arm < 2; arm++) {
-    for (int stn = 0; stn < NStationMAX; stn++) {
-      if (!StationStatus[stn])
+    for (int stn = 0; stn < kNStationMAX_; stn++) {
+      if (!stationStatus_[stn])
         continue;
-      for (int rp = 0; rp < NRPotsMAX; rp++) {
-        if (!RPstatus[stn][rp])
+      for (int rp = 0; rp < kNRPotsMAX_; rp++) {
+        if (!rpStatus_[stn][rp])
           continue;
         int index = getRPindex(arm, stn, rp);
-        if (RPindexValid[index] == 0)
+        if (RPindexValid_[index] == 0)
           continue;
 
-        for (int p = 0; p < NplaneMAX; p++) {
+        for (int p = 0; p < kNplaneMAX_; p++) {
           CTPPSPixelDetId planeId(arm, stn, rp, p);
-          if ((*pixDigi).find(planeId.rawId()) != (*pixDigi).end()) {
-            int n_digis = (*pixDigi)[planeId.rawId()].size();
-            h2HitsVsBXRandoms[index]->Fill(event.bunchCrossing(), p, n_digis);
+          auto pix_d = pixDigi->find(planeId.rawId());
+          if (pix_d != pixDigi->end()) {
+            int n_digis = pix_d->size();
+            h2HitsVsBXRandoms_[index]->Fill(event.bunchCrossing(), p, n_digis);
           }
         }
-      }  // end for (int rp=0; rp<NRPotsMAX; rp++) {
-    }    // end for (int stn = 0; stn < NStationMAX; stn++) {
+      }  // end for (int rp=0; rp<kNRPotsMAX_; rp++) {
+    }    // end for (int stn = 0; stn < kNStationMAX_; stn++) {
   }      // end for (int arm=0; arm<2; arm++) {
+}
+
+//---------------------------------------------------------------------------
+
+void CTPPSRandomDQMSource::fillDescriptions(edm::ConfigurationDescriptions &descriptions) {
+  edm::ParameterSetDescription desc;
+  desc.addUntracked<edm::InputTag>("tagRPixDigi", edm::InputTag("ctppsPixelDigisAlCaRecoProducer"));
+  desc.addUntracked<std::string>("folderName", "CTPPS/RandomPixel");
+  desc.addUntracked<unsigned int>("RPStatusWord", 0x8008);
+  descriptions.add("ctppsRandomDQMSource", desc);
 }
 
 //---------------------------------------------------------------------------

--- a/DQM/CTPPS/plugins/CTPPSRandomDQMSource.cc
+++ b/DQM/CTPPS/plugins/CTPPSRandomDQMSource.cc
@@ -78,7 +78,7 @@ private:
 //-------------------------------------------------------------------------------
 
 CTPPSRandomDQMSource::CTPPSRandomDQMSource(const edm::ParameterSet &ps)
-    : tokenDigi_(consumes<edm::DetSetVector<CTPPSPixelDigi>>(ps.getUntrackedParameter<edm::InputTag>("tagRPixDigi"))),
+    : tokenDigi_(consumes<edm::DetSetVector<CTPPSPixelDigi>>(ps.getParameter<edm::InputTag>("tagRPixDigi"))),
       folderName_(ps.getUntrackedParameter<std::string>("folderName", "CTPPS/RandomPixel")),
       rpStatusWord_(ps.getUntrackedParameter<unsigned int>("RPStatusWord", 0x8008)) {
   for (int stn = 0; stn < kStationIDMAX_; stn++) {
@@ -195,7 +195,7 @@ void CTPPSRandomDQMSource::analyze(edm::Event const &event, edm::EventSetup cons
 
 void CTPPSRandomDQMSource::fillDescriptions(edm::ConfigurationDescriptions &descriptions) {
   edm::ParameterSetDescription desc;
-  desc.addUntracked<edm::InputTag>("tagRPixDigi", edm::InputTag("ctppsPixelDigisAlCaRecoProducer"));
+  desc.add<edm::InputTag>("tagRPixDigi", edm::InputTag("ctppsPixelDigisAlCaRecoProducer"));
   desc.addUntracked<std::string>("folderName", "CTPPS/RandomPixel");
   desc.addUntracked<unsigned int>("RPStatusWord", 0x8008);
   descriptions.add("ctppsRandomDQMSource", desc);

--- a/DQM/CTPPS/python/ctppsRandomDQMSource_cfi.py
+++ b/DQM/CTPPS/python/ctppsRandomDQMSource_cfi.py
@@ -4,5 +4,6 @@ from DQMServices.Core.DQMEDAnalyzer import DQMEDAnalyzer
 
 ctppsRandomDQMSource = DQMEDAnalyzer('CTPPSRandomDQMSource',
     tagRPixDigi = cms.untracked.InputTag("ctppsPixelDigisAlCaRecoProducer", ""),
-    RPStatusWord = cms.untracked.uint32(0x8008), # rpots in readout:220_fr_hr; 210_fr_hr
+    folderName = cms.untracked.string("CTPPS/RandomPixel"),
+    RPStatusWord = cms.untracked.uint32(0x8008) # rpots in readout:220_fr_hr; 210_fr_hr
 )

--- a/DQM/CTPPS/python/ctppsRandomDQMSource_cfi.py
+++ b/DQM/CTPPS/python/ctppsRandomDQMSource_cfi.py
@@ -3,7 +3,7 @@ import FWCore.ParameterSet.Config as cms
 from DQMServices.Core.DQMEDAnalyzer import DQMEDAnalyzer
 
 ctppsRandomDQMSource = DQMEDAnalyzer('CTPPSRandomDQMSource',
-    tagRPixDigi = cms.untracked.InputTag("ctppsPixelDigisAlCaRecoProducer", ""),
+    tagRPixDigi = cms.InputTag("ctppsPixelDigisAlCaRecoProducer", ""),
     folderName = cms.untracked.string("CTPPS/RandomPixel"),
     RPStatusWord = cms.untracked.uint32(0x8008) # rpots in readout:220_fr_hr; 210_fr_hr
 )

